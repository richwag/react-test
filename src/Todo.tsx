import { useState } from "react";

export interface ITodo {
    userId: number;
    id: number;
    title: string;
    completed: boolean;
}

export function Todo({
    todo,
    onSelected,
}: {
    todo: ITodo;
    onSelected: Function;
}) {
    const [selected, setSelected] = useState(false);

    function handleOnClick() {
        let newValue = !selected;

        setSelected(newValue);

        onSelected(todo, newValue);
    }

    return (
        <div
            className={"todo" + (selected ? " selected" : "")}
            onClick={() => handleOnClick()}
        >
            <div className="todoTitle">{todo.title}</div>
            <div className="todoStatus">
                {todo.completed ? "complete" : "incomplete"}{" "}
            </div>
        </div>
    );
}
