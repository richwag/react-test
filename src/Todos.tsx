import { useEffect, useState } from "react";
import { Todo, ITodo } from "./Todo";

function useGetTodos(): [ITodo[] | null, Error | null] {
    const [todos, setTodos] = useState<ITodo[] | null>(null);
    const [error, setError] = useState<Error | null>(null);

    useEffect(() => {
        async function getTodos() {
            try {
                const response = await fetch(
                    "https://jsonplaceholder.typicode.com/todos"
                );

                setError(null);

                if (!response.ok) {
                    throw new Error("Unable to return");
                }

                const resp = await response.json();

                setTodos(resp);
            } catch (e: any) {
                setError(e);
                setTodos([]);
            }
        }

        getTodos();
    }, []);

    return [todos, error];
}

export function Todos() {
    const [todos, error] = useGetTodos();

    function onSelected(todo: ITodo, selected: boolean) {

    }

    return (
        <div>
            {todos && todos.map((todo) => <Todo key={todo.id} todo={todo} onSelected={onSelected} />)}
            {error && error.message}
        </div>
    );
}
