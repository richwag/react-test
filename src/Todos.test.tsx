import { triggerAsyncId } from "async_hooks";
import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { Todos } from "./Todos";
import { ITodo } from "./Todo";

import fetchMock from "jest-fetch-mock";

fetchMock.enableMocks();

let container: Element | null = null;

beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
    fetchMock.resetMocks();
});

afterEach(() => {
    // cleanup on exiting
    if (container !== null) {
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    }

    fetchMock.resetMocks();
});

describe("todos", () => {
    it("renders", async () => {
        const fakeTodos: ITodo[] = [
            {
                id: 1,
                completed: true,
                title: "test",
                userId: 1,
            },
        ];

        fetchMock.mockResponseOnce(JSON.stringify(fakeTodos));

        await act(async () => {
            render(<Todos />, container);
        });

        let todoElement = container?.querySelector(".todo");

        expect(todoElement).toBeTruthy();

        expect(container).toMatchSnapshot();
    });

    it("renders error when retrieve fails", async () => {
        fetchMock.mockRejectOnce(new Error("Error"));

        // Use the asynchronous version of act to apply resolved promises
        await act(async () => {
            render(<Todos />, container);
        });

        expect(container).toMatchSnapshot();
    });
});
