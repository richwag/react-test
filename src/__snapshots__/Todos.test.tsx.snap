// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos renders 1`] = `
<div>
  <div>
    <div
      class="todo"
    >
      <div
        class="todoTitle"
      >
        test
      </div>
      <div
        class="todoStatus"
      >
        complete
         
      </div>
    </div>
  </div>
</div>
`;

exports[`todos renders error when retrieve fails 1`] = `
<div>
  <div>
    Error
  </div>
</div>
`;
